src/Middleware/HttpLogging/test/HttpLoggingMiddlewareTests.cs

```diff
@@ -0,0 +1,757 @@
+                 },
+                 options,
+                 LoggerFactory);
+
+            var httpContext = new DefaultHttpContext();
+
+            // Header on the default allow list.
+            httpContext.Request.Headers["Connection"] = "keep-alive";
+
+            httpContext.Request.Headers["foo"] = "bar";
+
+            await middleware.Invoke(httpContext);
+            Assert.Contains(TestSink.Writes, w => w.Message.Contains("foo: bar"));
+            Assert.Contains(TestSink.Writes, w => w.Message.Contains("Connection: X"));
+        }
+
+        [Theory]
+        [MemberData(nameof(BodyData))]
+        public async Task RequestBodyReadingWorks(string expected)
+        {
+            var options = CreateOptionsAccessor();
+            options.CurrentValue.LoggingFields = HttpLoggingFields.RequestBody;
+
+            var middleware = new HttpLoggingMiddleware(
+                async c =>
+                {
+                    var arr = new byte[4096];
+                    while (true)
+                    {
+                        var res = await c.Request.Body.ReadAsync(arr);
+                        if (res == 0)
+                        {
+                            break;
+                        }
+                    }
+                },
+                options,
+                LoggerFactory);
+
+            var httpContext = new DefaultHttpContext();
+            httpContext.Request.ContentType = "text/plain";
+            httpContext.Request.Body = new MemoryStream(Encoding.UTF8.GetBytes(expected));
+
+            await middleware.Invoke(httpContext);
+
+            Assert.Contains(TestSink.Writes, w => w.Message.Contains(expected));
+        }
+
+        [Fact]
+        public async Task RequestBodyReadingLimitWorks()
+        {
+            var input = string.Concat(new string('a', 30000), new string('b', 3000));
+            var options = CreateOptionsAccessor();
+            options.CurrentValue.LoggingFields = HttpLoggingFields.RequestBody;
+
+            var middleware = new HttpLoggingMiddleware(
+                async c =>
+                {
+                    var arr = new byte[4096];
+                    while (true)
+                    {
+                        var res = await c.Request.Body.ReadAsync(arr);
+                        if (res == 0)
+                        {
+                            break;
+                        }
+                    }
+                },
+                options,
+                LoggerFactory);
+
+            var httpContext = new DefaultHttpContext();
+            httpContext.Request.ContentType = "text/plain";
+            httpContext.Request.Body = new MemoryStream(Encoding.UTF8.GetBytes(input));
+
+            await middleware.Invoke(httpContext);
+            var expected = input.Substring(0, options.CurrentValue.ResponseBodyLogLimit);
+
+            Assert.Contains(TestSink.Writes, w => w.Message.Contains(expected));
```

Comment:
I don't think this is checking the limit. `w.Message.Contains(expected)` will do a partial string match right?

Reply:
I realized this, I guess I have to check for new line as well?